lib/scholar/metrics.ex

```diff
@@ -369,6 +369,93 @@ defmodule Scholar.Metrics do
     Nx.indexed_add(zeros, indices, updates)
   end
 
+  @doc ~S"""
+  Calculates F1 score given rank-1 tensors which represent
+  the expected (`y_true`) and predicted (`y_pred`) classes.
+
+  ## Options
+
+    * `:num_classes` - required. Number of classes contained in the input tensors
+    * `:average` - optional. This determines the type of averaging performed on the data.
+      * `nil`. Default value. In this case, the scores for each class are returned.
+      * `:macro`. Calculate metrics for each label, and find their unweighted mean.
+        This does not take label imbalance into account.
+      * `:weighted`. Calculate metrics for each label, and find their average weighted by
+        support (the number of true instances for each label).
+      * `:micro`. Calculate metrics globally by counting the total true positives,
+        false negatives and false positives.
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 1, 1, 1, 1, 0, 2, 1, 0, 1], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 2, 1, 1, 2, 2, 2, 0, 0, 1], type: {:u, 32})
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: nil)
+      #Nx.Tensor<
+        f32[3]
+        [0.6666666865348816, 0.6666666865348816, 0.4000000059604645]
+      >
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: :macro)
+      #Nx.Tensor<
+        f32
+        0.5777778029441833
+      >
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: :weighted)
+      #Nx.Tensor<
+        f32
+        0.64000004529953
+      >
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: :micro)
+      #Nx.Tensor<
+        f32
+        0.6000000238418579
+      >
+  """
+  defn f1_score(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, [:num_classes, :average])
+
+    assert_shape_pattern(y_true, {_})
+    assert_shape(y_pred, Nx.shape(y_true))
+
+    num_classes =
+      transform(opts[:num_classes], fn num_classes ->
+        num_classes || raise ArgumentError, "missing option :num_classes"
+      end)
+
+    average =
+      transform(opts[:average], fn average ->
+        average || nil
+      end)
+
+    cm = confusion_matrix(y_true, y_pred, num_classes: num_classes)
+    true_positive = Nx.take_diagonal(cm)
+    false_positive = Nx.subtract(Nx.sum(cm, axes: [0]), true_positive)
+    false_negative = Nx.subtract(Nx.sum(cm, axes: [1]), true_positive)
+    precision = Nx.divide(true_positive, Nx.add(true_positive, false_positive))
+    recall = Nx.divide(true_positive, Nx.add(true_positive, false_negative))
```

Comment:
Aren't there functions to ger precision, recall and so on already?

Reply:
Existing functions only works for binary case (the label is either 0 or 1). The code here works for any number of classes.

Reply:
Shouldn't they work for other cases as well?
Perhaps it should ne changed as part of a separate PR on which this one depends

Reply:
I think that could be changed in a separate PR, however, the approach here would require to calculate the confusion matrix first (which should be pretty fast for binary labels). If that's okay, then I can send a separate PR. :)

Reply:
> the approach here would require to calculate the confusion matrix first (which should be pretty fast for binary labels)

We can use a different implementation when `num_classes: 2`.

The primary difference is that the current implementations accept a `:threshold`, and so they are semantically different. But maybe it's better to apply such normalization (probabilities -> classes) before using the metrics. I believe the implementations come from Axon, @seanmor5 thoughts?

Reply:
Another thing to note, `0` stands for `negative` in binary-class cases, however, in multi-class cases `0` should be treated as any other label. 

In simple words, even if `num_classes` is 2, in binary cases `1` and `0` represent `true` and `false` respectively, whereas the multi-class cases `1` and `0` could represent `cat` and `dog`. 

Therefore, when considering true positives, we only count the number of samples whose y_pred == 1 and y_true == 1. However, for multi-class, true positives should be the sum of samples whose (y_pred == 0 and y_true == 0) **and** (y_pred == 1 and y_true == 1).

Reply:
Say we have

```
y_true = Nx.tensor([1, 0, 1, 1, 0, 1, 0])
y_pred = Nx.tensor([1, 1, 0, 0, 1, 0, 0])
```

then the number of true positives should be 1 in binary-class case, while it should be 2 in multi-class case.

Binary-class:
```
y_true = Nx.tensor([1, 0, 1, 1, 0, 1, 0])
                    ^
y_pred = Nx.tensor([1, 1, 0, 0, 1, 0, 0])
                    ^
```

Multi-class:
```
y_true = Nx.tensor([1, 0, 1, 1, 0, 1, 0])
                    ^                 ^
y_pred = Nx.tensor([1, 1, 0, 0, 1, 0, 0])
                    ^                 ^
```

Reply:
Maybe we should set `num_classes` to 1 to denote the former case (1/0 denotes true/false). This should make sense because the question for the former case could be 'is this a cat' (only 1 class), while the latter case is asking 'is this cat or dog' (2 classes).

Reply:
Looking at [sklearn.metrics.recall_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.recall_score.html), they support `average="binary"` and return the value only for the true class (configurable with `pos_label`). For binary input they make it the default behaviour:

```
>>> y_true = [1, 0, 1, 1, 0, 1, 0]
>>> y_pred = [1, 1, 0, 0, 1, 0, 0]
>>> recall_score(y_true, y_pred)
0.25
>>> recall_score(y_true, y_pred, average=None)
array([0.33333333, 0.25      ])
```

Reply:
yeah, that's true. I think this also depends on the perspective, if we use the One-vs-Rest strategies for multi-class cases, then binary interpretation seems to be more general.

Reply:
It seems like the binary-class version is semantically a separate function. Again, looking at sklearn, the `:average` option doesn't make sense for binary (maybe that's why they use `average="binary"`, which reads pretty weird to be honest). On the other hand, the `pos_label` option doesn't make sense for multi-class.

There is a huge number of metrics, so I can see striving the decision to prefer a single parameterizable function, but at the same time, having `binary_recall_score` and `recall_score` may just be less confusing, especially that we want to special case the implementation for them anyway.

I think `num_classes: 1` can be confusing, an alternative approach would be `num_classes: :binary` (maybe no longer `num_`). Something along this lines sounds better than `average: :binary`, but I'm not convinced if that's the best we can do.

Could definitely use other opinions on this :D

(Sidenote: having confusion matrix, maybe we no longer need to expose the individual `tp` `fp` `tn` `tn` functions. sklearn doesn't)

Reply:
I like the idea of having two sets of functions too (one for binary-class and the other for multi-class). It seems to be less confusing and easier to maintain. Besides that, there's probably no way to combine them in a single function without some sort of if-statement inside.